### DISCUSSION TITLE
How to findPath int nested struct

### DISCUSSION THREAD

**Discussion by chenlinli** [cue-lang/cue#2053] [Q&A]
```
{
	"aa": 1,
	"fruits": [{
		"name": "a",
		"age": 13,
		"tool": [{
			"a": 1,
			"b": 2,
			"c": "{\"d\":1}"
		}]
	}, {
		"name": "bgggg",
		"add": "fgfgfjjjkj",
		"tool": [{
			"a": 1,
			"b": 2,
			"c": "{\"d\":1}"
		}]
	}]
}
```

For the given json before,how to find the field "c"'s all value by given path 'fruits.tool.c'? It seems that I should LookUpPath "fruits" first,then travel the fruits list , then LookUpPath "tool" and travel the list "tool",final LookUpPath "c" and get "c"'s value. But when change the json structure,and look for any path's value, it is very hard to do this in a common way. It seems that Go API can't tell the type of a field,like list,struct and so on

**Comment by myitcv** [comment]
@chenlinli the JSON you have provided does not appear to be valid:

https://cuelang.org/play/?id=tGvDKZU77G4#cue@export@cue

Perhaps you could try posting again?

**Reply by myitcv** [reply]
Please can you also use code blocks to better format code samples?

https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#quoting-code

**Reply by chenlinli** [reply]
`{
	"aa": 1,
	"fruits": [{
		"name": "a",
		"age": 13,
		"tool": [{
			"a": 1,
			"b": 2,
			"c": "{\"d\":1}"
		}]
	}, {
		"name": "bgggg",
		"add": "fgfgfjjjkj",
		"tool": [{
			"a": 1,
			"b": 2,
			"c": "{\"d\":1}"
		}]
	}]
}`
json

**Reply by chenlinli** [reply]
The json is just a case,and the structure is always change.my purpose is check an arbitrary json,and get any path of the json's value.

**Reply by myitcv** [reply]
FYI the easiest way to write a code block is like this:

    This is a test
    
    ```
    {
        "aa": 1,
        "fruits": [
            {
                "age": 13,
                "name": "a",
                "tool": [
                    {
                        "a": 1,
                        "b": 2,
                        "c": "{\"d\":1}"
                    }
                ]
            },
            {
                "add": "fgfgfjjjkj",
                "name": "bgggg",
                "tool": [
                    {
                        "a": 1,
                        "b": 2,
                        "c": "{\"d\":1}"
                    }
                ]
            }
        ]
    }
    ```

**Reply by myitcv** [reply]
Depending on what you are trying to achieve you can use [`cue.Value.Walk()`](https://pkg.go.dev/cuelang.org/go/cue#Value.Walk):

```
go mod tidy
go run .
cmp stdout stdout.golden

-- go.mod --
module mod.com

go 1.18

require cuelang.org/go v0.4.3

-- main.go --
package main

import (
	"fmt"

	"cuelang.org/go/cue"
	"cuelang.org/go/cue/cuecontext"
	"cuelang.org/go/cue/load"
)

func main() {
	ctx := cuecontext.New()
	bps := load.Instances([]string{"input.cue"}, nil)
	v := ctx.BuildInstance(bps[0])
	cPath := cue.ParsePath("c")
	v.Walk(func(x cue.Value) bool {
		if cVal := x.LookupPath(cPath); cVal.Exists() {
			fmt.Printf("path: %v, found c: %v\n", cVal.Path(), cVal)
			return false
		}
		return true
	}, nil)

}
-- input.cue --
{
    "aa": 1,
    "fruits": [
        {
            "age": 13,
            "name": "a",
            "tool": [
                {
                    "a": 1,
                    "b": 2,
                    "c": "{\"d\":1}"
                }
            ]
        },
        {
            "add": "fgfgfjjjkj",
            "name": "bgggg",
            "tool": [
                {
                    "a": 1,
                    "b": 2,
                    "c": "{\"d\":1}"
                }
            ]
        }
    ]
}
-- stdout.golden --
path: fruits[0].tool[0].c, found c: "{\"d\":1}"
path: fruits[1].tool[0].c, found c: "{\"d\":1}"
```

**Reply by chenlinli** [reply]
I tried your code,and it works.but there is a question that why the json content is named "input.cue" rather than "input.json". and when I renamed it to "input.json",it didn't work.

**Comment by myitcv** [comment]
> It seems that Go API can't tell the type of a field,like list,struct and so on

Please see https://pkg.go.dev/cuelang.org/go/cue#Value.Kind

**Reply by chenlinli** [reply]
I got it
